arch/riscv/core/pmp.c

```diff
@@ -0,0 +1,247 @@
+ * @param index_p Location of the PMP entry index to use. This index will be
+ *                updated according to the number of PMP slots needed
+ *                (even in the case of an overflow).
+ * @param perm PMP permission flags
+ * @param start Start address of the memory area to cover
+ * @param size Size of the memory area to cover
+ * @param pmp_addr Array of pmpaddr values (starting at entry 0).
+ * @param pmp_cfg Array of pmpcfg values (starting at entry 0).
+ * @param index_limit Index value representing the size of the provided arrays.
+ */
+static void set_pmp_entry(unsigned int *index_p, uint8_t perm,
+			  uintptr_t start, size_t size,
+			  ulong_t *pmp_addr, ulong_t *pmp_cfg,
+			  unsigned int index_limit)
+{
+	uint8_t *pmp_n_cfg = (uint8_t *)pmp_cfg;
+	unsigned int index = *index_p;
+
+	__ASSERT((start & 0x3) == 0, "misaligned start address");
+	__ASSERT((size & 0x3) == 0, "misaligned size");
+
+	if (index >= index_limit) {
+		LOG_ERR("out of PMP slots");
+		index += 1;
+	} else if ((index == 0 && start == 0) ||
+		   (index != 0 && pmp_addr[index - 1] == PMP_ADDR(start))) {
+		/* We can use TOR using only one additional slot */
+		pmp_addr[index] = PMP_ADDR(start + size);
+		pmp_n_cfg[index] = perm | PMP_TOR;
+		index += 1;
+	} else if (((size  & (size - 1)) == 0) /* power of 2 */ &&
+		   ((start & (size - 1)) == 0) /* naturally aligned */) {
+		pmp_addr[index] = PMP_ADDR_NAPOT(start, size);
+		pmp_n_cfg[index] = perm | (size == 4 ? PMP_NA4 : PMP_NAPOT);
+		index += 1;
+	} else if (index + 1 >= index_limit) {
+		LOG_ERR("out of PMP slots");
+		index += 2;
+	} else {
+		pmp_addr[index] = PMP_ADDR(start);
+		pmp_n_cfg[index] = 0;
+		index += 1;
+		pmp_addr[index] = PMP_ADDR(start + size);
+		pmp_n_cfg[index] = perm | PMP_TOR;
+		index += 1;
+	}
+	*index_p = index;
+}
+
+/**
+ * @brief Write a range of PMP entries to corresponding PMP registers
+ *
+ * PMP registers are accessed with the csr instruction which only takes an
+ * immediate value as the actual register. Efficiently accessing them is
+ * performed in pmp.S in the most efficient way possible.
+ *
+ * Requirement: start < end && end <= CONFIG_PMP_SLOT
+ *
+ * @param start Start of the PMP range to be written
+ * @param end End (exclusive) of the PMP range to be written
+ * @param pmp_addr Array of pmpaddr values (starting at entry 0).
+ * @param pmp_cfg Array of pmpcfg values (starting at entry 0).
+ * @param clear_trailing_entries True if trailing entries must be turned off
+ */
+extern void z_riscv_write_pmp_entries(unsigned int start, unsigned int end,
+				      ulong_t *pmp_addr, ulong_t *pmp_cfg,
+				      bool clear_trailing_entries);
+
+static void write_pmp_entries(unsigned int start, unsigned int end,
+			      ulong_t *pmp_addr, ulong_t *pmp_cfg,
+			      bool clear_trailing_entries)
+{
+	__ASSERT(start < end && end <= CONFIG_PMP_SLOT,
+		 "bad PMP range (start = %u end=%u)", start, end);
+
+	/* Be extra paranoid in case assertions are disabled */
```

Comment:
If for whatever reason the provided index is corrupted then this could go very wrong. Since this is security sensitive it's better to be defensive here.